### DISCUSSION TITLE
Always have to restart Z2M HA addon twice

### DISCUSSION THREAD

**Discussion by rikmueller** [Koenkk/zigbee2mqtt#15506] [General]
### What happend?
The Z2M HA addon won´t start after stopping it and restarting it. 
It will work after stopping and starting a second time.

### What did you expect to happen?
Restart should work every time

### Zigbee2MQTT version
1.28.4

### Adapter
Sonoff ZBDongle-E (EFR32MG21)

### Adapter firmware version
6.10.3.0 build 297

### HA version
2022.12.3

### Debug log
First restart;
`debug 2022-12-12 09:47:09: Loaded state from file /config/zigbee2mqtt/state.json
info  2022-12-12 09:47:09: Logging to console and directory: '/config/zigbee2mqtt/log/2022-12-12.09-47-09' filename: log.txt
debug 2022-12-12 09:47:09: Removing old log directory '/config/zigbee2mqtt/log/2022-12-11.15-55-01'
info  2022-12-12 09:47:09: Starting Zigbee2MQTT version 1.28.4 (commit #unknown)
info  2022-12-12 09:47:09: Starting zigbee-herdsman (0.14.76)
debug 2022-12-12 09:47:09: Using zigbee-herdsman with settings: '{"adapter":{"concurrent":null,"delay":null,"disableLED":false},"backupPath":"/config/zigbee2mqtt/coordinator_backup.json","databaseBackupPath":"/config/zigbee2mqtt/database.db.backup","databasePath":"/config/zigbee2mqtt/database.db","network":{"channelList":[15],"extendedPanID":[221,221,221,221,221,221,221,221],"networkKey":"HIDDEN","panID":4711},"serialPort":{"adapter":"ezsp","path":"/dev/ttyACM0"}}'`

Stop restart:
`error 2022-12-12 09:48:37: Failed to call 'Frontend' 'stop' (TypeError: Cannot read properties of null (reading 'clients')
    at Frontend.stop (/app/lib/extension/frontend.ts:59:39)
    at Controller.callExtensions (/app/lib/controller.ts:316:40)
    at Controller.stop (/app/lib/controller.ts:193:9)
    at stop (/app/index.js:111:5)
    at process.handleQuit (/app/index.js:117:9))`

**Comment by ttowntom** [comment]
+1
I have the same issue, only I'm using both HA and Z2M as dockers (not plugin).
The Zigbee2Mqtt docker needs a second restart for it to actually boot.

**Reply by Koenkk** [reply]
@ttowntom also with the same error?

**Reply by ttowntom** [reply]
@Koenkk these are the log outputs:

Start:
`debug 2023-01-02 10:08:19: Loaded state from file /app/data/state.json
info  2023-01-02 10:08:19: Logging to console and directory: '/app/data/log/2023-01-02.10-08-19' filename: log.txt
debug 2023-01-02 10:08:19: Removing old log directory '/app/data/log/2023-01-02.05-53-55'
info  2023-01-02 10:08:19: Starting Zigbee2MQTT version 1.29.0 (commit #dddce57)
info  2023-01-02 10:08:19: Starting zigbee-herdsman (0.14.83)
debug 2023-01-02 10:08:19: Using zigbee-herdsman with settings: '{"adapter":{"concurrent":null,"delay":null,"disableLED":false},"backupPath":"/app/data/coordinator_backup.json","databaseBackupPath":"/app/data/database.db.backup","databasePath":"/app/data/database.db","network":{"channelList":[25],"extendedPanID":[221,221,221,221,221,221,221,221],"networkKey":"HIDDEN","panID":6754},"serialPort":{"adapter":"ezsp","path":"/dev/serial/by-id/usb-ITEAD_SONOFF_Zigbee_3.0_USB_Dongle_Plus_V2_20220817082510-if00"}}'
`
This is where it "hangs".
I then restart it and the log writes these extra lines:
`error 2023-01-02 10:10:11: Failed to call 'Frontend' 'stop' (TypeError: Cannot read properties of null (reading 'clients')
    at Frontend.stop (/app/lib/extension/frontend.ts:83:39)
    at Controller.callExtensions (/app/lib/controller.ts:316:40)
    at Controller.stop (/app/lib/controller.ts:193:9)
    at stop (/app/index.js:111:5)
    at process.handleQuit (/app/index.js:117:9))`

Restart:
`debug 2023-01-02 10:10:19: Loaded state from file /app/data/state.json
info  2023-01-02 10:10:19: Logging to console and directory: '/app/data/log/2023-01-02.10-10-19' filename: log.txt
debug 2023-01-02 10:10:19: Removing old log directory '/app/data/log/2023-01-02.08-29-57'
info  2023-01-02 10:10:19: Starting Zigbee2MQTT version 1.29.0 (commit #dddce57)
info  2023-01-02 10:10:19: Starting zigbee-herdsman (0.14.83)
debug 2023-01-02 10:10:19: Using zigbee-herdsman with settings: '{"adapter":{"concurrent":null,"delay":null,"disableLED":false},"backupPath":"/app/data/coordinator_backup.json","databaseBackupPath":"/app/data/database.db.backup","databasePath":"/app/data/database.db","network":{"channelList":[25],"extendedPanID":[221,221,221,221,221,221,221,221],"networkKey":"HIDDEN","panID":6754},"serialPort":{"adapter":"ezsp","path":"/dev/serial/by-id/usb-ITEAD_SONOFF_Zigbee_3.0_USB_Dongle_Plus_V2_20220817082510-if00"}}'
info  2023-01-02 10:10:24: zigbee-herdsman started (resumed)
info  2023-01-02 10:10:24: Coordinator firmware version: '{"meta":{"maintrel":"3 ","majorrel":"6","minorrel":"10","product":8,"revision":"6.10.3.0 build 297"},"type":"EZSP v8"}'
debug 2023-01-02 10:10:24: Zigbee network parameters: {"channel":25,"extendedPanID":221,"panID":6754}
info  2023-01-02 10:10:24: Currently 25 devices are joined:`
..and aaaall the other lines indicating the devices and their states.
`info  2023-01-02 10:10:24: Connecting to MQTT server at mqtt://10.0.1.19
debug 2023-01-02 10:10:24: Using MQTT login with username: ttowntom
info  2023-01-02 10:10:24: Connected to MQTT server`
And then it's up and running as normal.

**Reply by Koenkk** [reply]
It looks like the EZSP startup is hanging. Can you provide me the **herdsman** debug log?

See https://www.zigbee2mqtt.io/guide/usage/debug.html on how to enable the herdsman debug logging. Note that this is only logged to STDOUT and not to log files.

**Reply by ttowntom** [reply]
Hi @Koenkk !
I've attached the log file as it is loooong - even after a cleanup. 
Where the file now begins (timestamp 10:50) is when the docker is started, and hangs. At timestamp 10:51 the docker is restarted and consecvently starts normally.
[ttowntomZ2m.log](https://github.com/Koenkk/zigbee2mqtt/files/10336009/ttowntomZ2m.log)

_I run docker on Unraid, and it's trying to make docker management easier, however that means it's also not doing docker management the "standard way". For instance, I can't write docker run files, and the dokcer logs are hidden away (and did not seem to quite follow the `docker logs ZIGBEE2MQTT_CONTAINER_NAME > log.txt 2>&1` scheme).  So please let me now if this does not look like what you need :)_

**Reply by Koenkk** [reply]
@kirovilya it seems that the ezsp init fails, relevant logging:

```
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart \u003c-- [ce4b83fc7d5e1644e3c90a6a7ec20251a8bd7e]\n","stream":"stderr","time":"2023-01-03T10:50:09.435476037Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart \u003c-- DATA (0,7,0): 07d4b1a9112a15b658964e24ab1593499cabfb78c9809874f5ce4b83fc7e1644e3c90a6a7e\n","stream":"stderr","time":"2023-01-03T10:50:09.435542411Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart --\u003e ACK  (1)\n","stream":"stderr","time":"2023-01-03T10:50:09.435561617Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart --\u003e [8160597e]\n","stream":"stderr","time":"2023-01-03T10:50:09.435591222Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart Unexpected packet sequence 7 | 0\n","stream":"stderr","time":"2023-01-03T10:50:09.435624334Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:ezsp \u003c== Frame: 96900145000004010204010140010000e5dcd3244effff0808280a000029e30804\n","stream":"stderr","time":"2023-01-03T10:50:09.435771187Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:ezsp \u003c== 0x1: {\"_id_\":1,\"_cls_\":\"getLibraryStatus\",\"_isRequest_\":false,\"status\":69}\n","stream":"stderr","time":"2023-01-03T10:50:09.435806934Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart \u003c-- Error: c20251a8bd7e\n","stream":"stderr","time":"2023-01-03T10:50:09.435872095Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart Uart reseting\n","stream":"stderr","time":"2023-01-03T10:50:09.435893996Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart --\u003e Write reset\n","stream":"stderr","time":"2023-01-03T10:50:09.435910677Z"}
{"log":"2023-01-03T10:50:09.435Z zigbee-herdsman:adapter:ezsp:uart --\u003e [1ac038bc7e]\n","stream":"stderr","time":"2023-01-03T10:50:09.43600349Z"}
{"log":"2023-01-03T10:50:09.555Z zigbee-herdsman:adapter:ezsp:uart \u003c-- [11]\n","stream":"stderr","time":"2023-01-03T10:50:09.555668365Z"}
{"log":"2023-01-03T10:50:10.555Z zigbee-herdsman:adapter:ezsp:uart \u003c-- [c1020b0a527e]\n","stream":"stderr","time":"2023-01-03T10:50:10.555364533Z"}
{"log":"2023-01-03T10:50:10.555Z zigbee-herdsman:adapter:ezsp:uart \u003c-- RSTACK: c1020b0a527e\n","stream":"stderr","time":"2023-01-03T10:50:10.555542314Z"}
{"log":"2023-01-03T10:50:10.555Z zigbee-herdsman:adapter:ezsp:uart RSTACK Version: 2 Reason: RESET_SOFTWARE frame: c1020b0a527e\n","stream":"stderr","time":"2023-01-03T10:50:10.556281793Z"}
{"log":"Zigbee2MQTT:error 2023-01-03 11:51:04: Failed to call 'Frontend' 'stop' (TypeError: Cannot read properties of null (reading 'clients')\n","stream":"stdout","time":"2023-01-03T10:51:04.600301452Z"}
{"log":"    at Frontend.stop (/app/lib/extension/frontend.ts:83:39)\n","stream":"stdout","time":"2023-01-03T10:51:04.600338231Z"}
{"log":"    at Controller.callExtensions (/app/lib/controller.ts:316:40)\n","stream":"stdout","time":"2023-01-03T10:51:04.600346496Z"}
{"log":"    at Controller.stop (/app/lib/controller.ts:193:9)\n","stream":"stdout","time":"2023-01-03T10:51:04.600352427Z"}
{"log":"    at stop (/app/index.js:111:5)\n","stream":"stdout","time":"2023-01-03T10:51:04.600358789Z"}
```

**Reply by ttowntom** [reply]
Yup.
From what I understand it's not a "stable" driver..? Need it to run the adapter, though.
For now I guess double-starting z2m is the way to go.

**Reply by Koenkk** [reply]
@ttowntom EZSP support is indeed in an experimental phase.

**Reply by tsspmq** [reply]
Yep, I switched from EZSP (Nortek one) to the Sonoff 3.0 Plus yesterday and the hang ups, devices just leaving the network, map timing out and never being complete.  It's 100x better with the TI based controller and now I have a spare stick to experiment with.   Win/Win.   Another thing that annoyed me with ezsp was LQI was 255 on everything, even if it was in a weak location.   Once the new stick came in started seeing what I consider accurate numbers for those few devices that need help.   

All of those issues were also present in the ZHA integration, not just Z2M Edge except the double starts.  

Hanging up on start was about the only unique issue to Z2M, and it was once the network got very large.

I really didnt want to switch sticks but already had a Zooz 700 Z-Wave stick, so didnt need the Nortek side anymore.  Best $20 I've spent in a while.

**Comment by ttowntom** [comment]
For anyone else with the same issue:

I wrote a simple bash script to check if zigbee2mqtt reponds to http requests. If it does not the driver most likely hangs on boot, and thus the scripts restarts the docker. I have it run as a cron job every hour on my server.

```bash
#!/bin/bash

status_code=$(curl -s -o /dev/null -w "%{http_code}" http://insert-your-z2m-ip-and-port-here)

if [[ $status_code -eq 200 ]]
then
    :
else
    docker restart zigbee2mqtt
fi
```
